api/routes_payment.go

```diff
@@ -1,69 +1,317 @@
+		api.LogError(err, eh.QueueInitializationError)(c, http.StatusBadRequest)
+		return
+	}
+	if err = qm.PublishMessage(paymentConfirmation); err != nil {
+		api.LogError(err, eh.QueuePublishError)(c, http.StatusBadRequest)
+		return
+	}
+	Respond(c, http.StatusOK, gin.H{"response": payment})
+}
+
+// RequestSignedPaymentMessage is used to get a signed message from the GRPC API Payments Server
+func (api *API) RequestSignedPaymentMessage(c *gin.Context) {
 	username := GetAuthenticatedUserFromContext(c)
 	paymentType, exists := c.GetPostForm("payment_type")
 	if !exists {
 		FailWithMissingField(c, "payment_type")
 		return
 	}
-	usdValue, err := api.getUSDValue(paymentType)
+	var method uint64
+	switch paymentType {
+	case "0":
+		paymentType = "rtc"
+		method = 0
+	case "1":
+		paymentType = "eth"
+		method = 1
+	default:
+		Fail(c, errors.New("payment_type must be '0 (rtc)' or '1 (eth)'"))
+		return
+	}
+	senderAddress, exists := c.GetPostForm("sender_address")
+	if !exists {
+		FailWithMissingField(c, "sender_address")
+		return
+	}
+	creditValue, exists := c.GetPostForm("credit_value")
+	if !exists {
+		FailWithMissingField(c, "credit_value")
+		return
+	}
+	usdValueFloat, err := api.getUSDValue(paymentType)
 	if err != nil {
 		api.LogError(err, eh.CmcCheckError)(c, http.StatusBadRequest)
 		return
 	}
-	depositAddress, err := api.getDepositAddress(paymentType)
+	pm := models.NewPaymentManager(api.dbm.DB)
+	paymentNumber, err := pm.GetLatestPaymentNumber(username)
+	if err != nil && err != gorm.ErrRecordNotFound {
+		api.LogError(err, eh.PaymentSearchError)(c, http.StatusBadRequest)
+		return
+	}
+	creditValueFloat, err := strconv.ParseFloat(creditValue, 64)
 	if err != nil {
-		api.LogError(err, eh.DepositAddressCheckError)(c, http.StatusBadRequest)
+		Fail(c, err)
 		return
 	}
-	txHash, exists := c.GetPostForm("tx_hash")
+	chargeAmountFloat := creditValueFloat / usdValueFloat
+	chargeAmountBig := utils.FloatToBigInt(chargeAmountFloat)
+	chargeAmountString := chargeAmountBig.String()
+	numberString := strconv.FormatInt(paymentNumber, 10)
+	methodString := strconv.FormatUint(method, 10)
+	signRequest := greq.SignRequest{
+		Address:      senderAddress,
+		Method:       methodString,
+		Number:       numberString,
+		ChargeAmount: chargeAmountString,
+	}
+	gc, err := gapi.NewGAPIClient(api.cfg, true)
+	if err != nil {
+		api.LogError(err, err.Error())(c, http.StatusBadRequest)
+		return
+	}
+	defer gc.GC.Close()
+	resp, err := gc.GetSignedMessage(context.Background(), &signRequest)
+	if err != nil {
+		api.LogError(err, err.Error())(c, http.StatusBadRequest)
+		return
+	}
+	paymentNumberString := fmt.Sprintf("%s-%s", username, strconv.FormatInt(paymentNumber, 10))
+
+	if _, err = pm.NewPayment(
+		paymentNumber,
+		paymentNumberString,
+		paymentNumberString,
+		creditValueFloat,
+		"ethereum",
+		paymentType,
+		username,
+	); err != nil {
+		api.LogError(err, err.Error())(c, http.StatusBadRequest)
+		return
+	}
+	hEncoded := resp.GetH()
```

Comment:
what's all this stuff doing? it might be cleaner to move this out of the endpoint - IMO we should try to limit the amount of logic in the endpoints